<h1> Probability distribution for each F1 driver to win the Singporean Grand Prix </h1>

In [8]:
#Probability Distribution function
class ProbDist(dict):
    """A Probability Distribution; an {outcome: probability} mapping."""
    def __init__(self, mapping=(), **kwargs):
        self.update(mapping, **kwargs)
        # Make probabilities sum to 1.0; assert no negative probabilities
        total = sum(self.values())
        for outcome in self:
            self[outcome] = self[outcome] / total
            assert self[outcome] >= 0

In [9]:
#Function defining p(event,space)
def p(event, space): 
    """The probability of an event, given a sample space of equiprobable outcomes. 
    event: a collection of outcomes, or a predicate that is true of outcomes in the event. 
    space: a set of outcomes or a probability distribution of {outcome: frequency} pairs."""
    if is_predicate(event):
        event = such_that(event, space)
    if isinstance(space, ProbDist):
        return sum(space[o] for o in space if o in event)
    else:
        return Fraction(len(event & space), len(space))

is_predicate = callable

#such_that() function
def such_that(predicate, space): 
    """The outcomes in the sample pace for which the predicate is true.
    If space is a set, return a subset {outcome,...} with outcomes where predicate(element) is true;
    if space is a ProbDist, return a ProbDist {outcome: frequency,...} with outcomes where predicate(element) is true."""
    if isinstance(space, ProbDist):
        return ProbDist({o:space[o] for o in space if predicate(o)})
    else:
        return {o for o in space if predicate(o)}

In [10]:
SGP = ProbDist(LH = 413, VB = 326, MV = 278, CL=264, SV=240, CS=96, PG=95, AA=92, DR=54, SP=52, LN=49, KR=43, DK=37,NH=37, LS=21, KM=20, AG=14, RG=8, RK=1, GR=0)
SGP

{'LH': 0.1929906542056075,
 'VB': 0.15233644859813084,
 'MV': 0.12990654205607477,
 'CL': 0.1233644859813084,
 'SV': 0.11214953271028037,
 'CS': 0.044859813084112146,
 'PG': 0.04439252336448598,
 'AA': 0.04299065420560748,
 'DR': 0.025233644859813085,
 'SP': 0.024299065420560748,
 'LN': 0.022897196261682243,
 'KR': 0.020093457943925235,
 'DK': 0.017289719626168223,
 'NH': 0.017289719626168223,
 'LS': 0.009813084112149532,
 'KM': 0.009345794392523364,
 'AG': 0.0065420560747663555,
 'RG': 0.003738317757009346,
 'RK': 0.00046728971962616824,
 'GR': 0.0}

This gives probabilty for each driver winning Singporean Grand Prix

<h1> Probability distribution for each F1 driver to win both Singporean and Russian Grand Mix </h1>

In [11]:
SGP = ProbDist(LH = 413, VB = 326, MV = 278, CL=264, SV=240, CS=96, PG=95, AA=92, DR=54, SP=52, LN=49, KR=43, DK=37,
              NH=37, LS=21, KM=20, AG=14, RG=8, RK=1, GR=0)
RGP = ProbDist(LH = 413, VB = 326, MV = 278, CL=264, SV=240, CS=96, PG=95, AA=92, DR=54, SP=52, LN=49, KR=43, DK=37,
              NH=37, LS=21, KM=20, AG=14, RG=8, RK=1, GR=0)

In [13]:
def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b]
                    for a in A
                    for b in B}) 

race = joint(SGP, RGP, ' ')
race

{'LH LH': 0.037245392610708354,
 'LH VB': 0.02939951087431216,
 'LH MV': 0.025070748536990126,
 'LH CL': 0.023808192855271196,
 'LH SV': 0.02164381168661018,
 'LH CS': 0.008657524674644071,
 'LH PG': 0.008567342125949864,
 'LH AA': 0.008296794479867236,
 'LH DR': 0.004869857629487291,
 'LH SP': 0.004689492532098873,
 'LH LN': 0.004418944886016246,
 'LH KR': 0.003877849593850991,
 'LH DK': 0.003336754301685736,
 'LH NH': 0.003336754301685736,
 'LH LS': 0.0018938335225783906,
 'LH KM': 0.0018036509738841816,
 'LH AG': 0.0012625556817189272,
 'LH RG': 0.0007214603895536728,
 'LH RK': 9.01825486942091e-05,
 'LH GR': 0.0,
 'VB LH': 0.02939951087431216,
 'VB VB': 0.023206393571490954,
 'VB MV': 0.019789501266486154,
 'VB CL': 0.018792907677526418,
 'VB SV': 0.017084461525024016,
 'VB CS': 0.006833784610009605,
 'VB PG': 0.006762599353655339,
 'VB AA': 0.006549043584592539,
 'VB DR': 0.0038440038431304035,
 'VB SP': 0.0037016333304218697,
 'VB LN': 0.00348807756135907,
 'VB KR': 0.00306096602

In [14]:
def each_win(outcome): return outcome == 'LH LH' or outcome == 'VB VB' or outcome == 'MV MV' or outcome == 'CL CL' or outcome == 'SV SV' or outcome == 'CS CS' or outcome == 'PG PG' or outcome == 'AA AA' or outcome == 'DR DR' or outcome == 'SP SP' or outcome == 'LN LN' or outcome == 'KR KR' or outcome == 'DK DK' or outcome == 'NH NH' or outcome == 'LS LS' or outcome == 'CL CL' or outcome == 'SV SV' or outcome == 'CS CS' or outcome == 'PG PG' or outcome == 'AA AA' or outcome == 'DR DR' or outcome == 'SP SP' or outcome == 'LN LN' or outcome == 'LS LS' or outcome == 'KM KM' or outcome == 'AG AG' or outcome == 'RG RG' or outcome == 'RK RK' or outcome == 'GR GR' 
such_that(each_win, race)
#p(each_win, race)

{'LH LH': 0.3268607235934386,
 'VB VB': 0.20365629311666406,
 'MV MV': 0.14809903418672388,
 'CL CL': 0.13355817875210788,
 'SV SV': 0.110378660125709,
 'CS CS': 0.01766058562011344,
 'PG PG': 0.01729457304921048,
 'AA AA': 0.016219530890694465,
 'DR DR': 0.005587919668864019,
 'SP SP': 0.0051816648781235625,
 'LN LN': 0.004601027134753947,
 'KR KR': 0.0035432316418825688,
 'DK DK': 0.0026234094741683267,
 'NH NH': 0.0026234094741683267,
 'LS LS': 0.0008450866165874595,
 'KM KM': 0.0007665184730952015,
 'AG AG': 0.00037559405181664877,
 'RG RG': 0.00012264295569523224,
 'RK RK': 1.916296182738004e-06,
 'GR GR': 0.0}

The above probabilities show that each F1 drivers win both races

<h1> Probability distribution for Mercedes to win both the race </h1>

In [16]:
def mercedes(outcome): return outcome == 'LH LH' or outcome == 'VB VB' or outcome == 'LH VB' or outcome == 'VB LH'
such_that(mercedes, race)

{'LH LH': 0.3123282203028267,
 'LH VB': 0.24653510851990676,
 'VB LH': 0.24653510851990676,
 'VB VB': 0.1946015626573598}

In [17]:
def mercedes(outcome): return outcome == 'LH LH' or outcome == 'VB VB' or outcome == 'LH VB' or outcome == 'VB LH'
p(mercedes, race)

0.11925080793082363

0.1192 is the probabilty that Mercedes wins both the race

<h1> Probability distribution of Mercedes winning at least one race </h1>

In [19]:
def mercedes(outcome): return 'LH' in outcome or 'VB' in outcome
such_that(mercedes, race)
p(mercedes, race)

0.5714033976766529

0.571 is probability of Mercedes winning at least one race

 <h1> Probability of Mercedes winning second race,after winning the first one </h1>

In [31]:
def starts(outcome):  return outcome.startswith('LH') or outcome.startswith('VB')
def ends(outcome): return outcome.endswith('LH') or outcome.endswith('VB')

In [32]:
p(ends, such_that(starts, race))

0.3453271028037384

0.345 is the probability of Mercedes winning second race after the first one

<h1> If Mercedes wins at least one of two races, probability that Mercedes wins both the races </h1>

In [38]:
def mercedes(outcome): return outcome == 'LH LH' or outcome == 'VB VB' or outcome == 'LH VB' or outcome == 'VB LH'
p(mercedes, race)

0.11925080793082363

0.1192 is the probabilty that Mercedes wins both the race

<h1> Probability of Ferrari winning second race,after winning the first one </h1>

In [33]:
def starts(outcome):  return outcome.startswith('CL') or outcome.startswith('SV')
def ends(outcome): return outcome.endswith('CL') or outcome.endswith('SV')

In [34]:
p(ends, such_that(starts, race))

0.23551401869158872

0.235 is the probability of Ferrari winning second race after the first one

<h1> If Ferrari wins at least one of two races, probability that Ferrari wins both the races </h1>

In [40]:
def ferrari(outcome): return outcome == 'CL CL' or outcome == 'SV SV' or outcome == 'CL SV' or outcome == 'SV CL'
p(ferrari, race)

0.05546685300026201

0.0554 is the probabilty that Mercedes wins both the race

<h1> Probability of Alpha winning second race,after winning the first one </h1>

In [36]:
def starts(outcome):  return outcome.startswith('KR') or outcome.startswith('AG')
def ends(outcome): return outcome.endswith('KG') or outcome.endswith('AG')

In [37]:
p(ends, such_that(starts, race))

0.0065420560747663555

0.00654 is the probability of Alpha winning second race

 <h1> If Alpha wins at least one of two races, probability that Alpha wins both the races </h1>

In [42]:
def alpha(outcome): return outcome == 'KR KR' or outcome == 'AG AG' or outcome == 'KR AG' or outcome == 'AG KR'
p(alpha, race)

0.0007094506070399161

0.00070 is the probability of winning both race

 <h1> Probability of McLaren winning second race,after winning the first one </h1>

In [44]:
def starts(outcome):  return outcome.startswith('CS') or outcome.startswith('LN')
def ends(outcome): return outcome.endswith('CS') or outcome.endswith('LN')

In [45]:
p(ends, such_that(starts, race))

0.06775700934579439

0.0677 is the probability of McLaren winning second race

<h1> If McLaren wins at least one of two races, probability that McLaren wins both the races </h1>

In [47]:
def mclaren(outcome): return outcome == 'CS CS' or outcome == 'LN LN' or outcome == 'CS LN' or outcome == 'LN CS'
p(mclaren, race)

0.004591012315486067

0.0045 is the probability of winning both races

<h1> Mercedes wins at least one of these two races on a rainy day. What is the probability Mercedes wins both races, assuming races can be held on either rainy, sunny, cloudy, snowy or foggy days? Assume that rain, sun, clouds, snow, and fog are the only possible weather conditions on race tracks. </h1>

In [48]:
#As there are 5 possibe weather conditions in a day, each weather condition will have 1/5 or 0.2 possibility
day = ProbDist(r = 0.2, sun = 0.2, c = 0.2, s = 0.2, f = 0.2)

In [49]:
def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b] 
                    for a in A
                    for b in B}) 

race1 = joint(SGP, day, ' ')
race1

{'LH r': 0.03859813084112147,
 'LH sun': 0.03859813084112147,
 'LH c': 0.03859813084112147,
 'LH s': 0.03859813084112147,
 'LH f': 0.03859813084112147,
 'VB r': 0.03046728971962615,
 'VB sun': 0.03046728971962615,
 'VB c': 0.03046728971962615,
 'VB s': 0.03046728971962615,
 'VB f': 0.03046728971962615,
 'MV r': 0.02598130841121494,
 'MV sun': 0.02598130841121494,
 'MV c': 0.02598130841121494,
 'MV s': 0.02598130841121494,
 'MV f': 0.02598130841121494,
 'CL r': 0.024672897196261666,
 'CL sun': 0.024672897196261666,
 'CL c': 0.024672897196261666,
 'CL s': 0.024672897196261666,
 'CL f': 0.024672897196261666,
 'SV r': 0.022429906542056063,
 'SV sun': 0.022429906542056063,
 'SV c': 0.022429906542056063,
 'SV s': 0.022429906542056063,
 'SV f': 0.022429906542056063,
 'CS r': 0.008971962616822425,
 'CS sun': 0.008971962616822425,
 'CS c': 0.008971962616822425,
 'CS s': 0.008971962616822425,
 'CS f': 0.008971962616822425,
 'PG r': 0.008878504672897192,
 'PG sun': 0.008878504672897192,
 'PG c': 

In [50]:
def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b] 
                    for a in A
                    for b in B}) 

race2 = joint(RGP, day, ' ')
race2

{'LH r': 0.03859813084112147,
 'LH sun': 0.03859813084112147,
 'LH c': 0.03859813084112147,
 'LH s': 0.03859813084112147,
 'LH f': 0.03859813084112147,
 'VB r': 0.03046728971962615,
 'VB sun': 0.03046728971962615,
 'VB c': 0.03046728971962615,
 'VB s': 0.03046728971962615,
 'VB f': 0.03046728971962615,
 'MV r': 0.02598130841121494,
 'MV sun': 0.02598130841121494,
 'MV c': 0.02598130841121494,
 'MV s': 0.02598130841121494,
 'MV f': 0.02598130841121494,
 'CL r': 0.024672897196261666,
 'CL sun': 0.024672897196261666,
 'CL c': 0.024672897196261666,
 'CL s': 0.024672897196261666,
 'CL f': 0.024672897196261666,
 'SV r': 0.022429906542056063,
 'SV sun': 0.022429906542056063,
 'SV c': 0.022429906542056063,
 'SV s': 0.022429906542056063,
 'SV f': 0.022429906542056063,
 'CS r': 0.008971962616822425,
 'CS sun': 0.008971962616822425,
 'CS c': 0.008971962616822425,
 'CS s': 0.008971962616822425,
 'CS f': 0.008971962616822425,
 'PG r': 0.008878504672897192,
 'PG sun': 0.008878504672897192,
 'PG c': 

In [51]:
def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b] 
                    for a in A
                    for b in B}) 

race = joint(race1, race2, ' ')
race

{'LH r LH r': 0.0014898157044283642,
 'LH r LH sun': 0.0014898157044283642,
 'LH r LH c': 0.0014898157044283642,
 'LH r LH s': 0.0014898157044283642,
 'LH r LH f': 0.0014898157044283642,
 'LH r VB r': 0.0011759804349725101,
 'LH r VB sun': 0.0011759804349725101,
 'LH r VB c': 0.0011759804349725101,
 'LH r VB s': 0.0011759804349725101,
 'LH r VB f': 0.0011759804349725101,
 'LH r MV r': 0.0010028299414796254,
 'LH r MV sun': 0.0010028299414796254,
 'LH r MV c': 0.0010028299414796254,
 'LH r MV s': 0.0010028299414796254,
 'LH r MV f': 0.0010028299414796254,
 'LH r CL r': 0.0009523277142108671,
 'LH r CL sun': 0.0009523277142108671,
 'LH r CL c': 0.0009523277142108671,
 'LH r CL s': 0.0009523277142108671,
 'LH r CL f': 0.0009523277142108671,
 'LH r SV r': 0.0008657524674644247,
 'LH r SV sun': 0.0008657524674644247,
 'LH r SV c': 0.0008657524674644247,
 'LH r SV s': 0.0008657524674644247,
 'LH r SV f': 0.0008657524674644247,
 'LH r CS r': 0.0003463009869857699,
 'LH r CS sun': 0.0003463009

In [66]:
def mercedes_rainy(outcome): 
    return outcome.startswith('LH r') == 1 or outcome.startswith('VB r') == 1 or outcome.endswith('LH r') == 1 or outcome.endswith('VB r') == 1

such_that(mercedes_rainy, race)

{'LH r LH r': 0.011171315754503128,
 'LH r LH sun': 0.011171315754503128,
 'LH r LH c': 0.011171315754503128,
 'LH r LH s': 0.011171315754503128,
 'LH r LH f': 0.011171315754503128,
 'LH r VB r': 0.008818036164571476,
 'LH r VB sun': 0.008818036164571476,
 'LH r VB c': 0.008818036164571476,
 'LH r VB s': 0.008818036164571476,
 'LH r VB f': 0.008818036164571476,
 'LH r MV r': 0.007519675011505738,
 'LH r MV sun': 0.007519675011505738,
 'LH r MV c': 0.007519675011505738,
 'LH r MV s': 0.007519675011505738,
 'LH r MV f': 0.007519675011505738,
 'LH r CL r': 0.007140986341861563,
 'LH r CL sun': 0.007140986341861563,
 'LH r CL c': 0.007140986341861563,
 'LH r CL s': 0.007140986341861563,
 'LH r CL f': 0.007140986341861563,
 'LH r SV r': 0.006491805765328694,
 'LH r SV sun': 0.006491805765328694,
 'LH r SV c': 0.006491805765328694,
 'LH r SV s': 0.006491805765328694,
 'LH r SV f': 0.006491805765328694,
 'LH r CS r': 0.0025967223061314775,
 'LH r CS sun': 0.0025967223061314775,
 'LH r CS c': 

In [67]:
def Mercedes_win(outcome):
    return outcome.count('LH') == 2 or outcome.count('VB') == 2 or outcome.count('LH') == 1 and outcome.count('VB') == 1
such_that(Mercedes_win, race)
p(Mercedes_win, such_that(mercedes_rainy, race))

0.32191084652243374

0.3219 is the probability of Mercedes winning any day given that it wins on a rainy day